velox/docs/develop/unsaferow.rst

```diff
@@ -0,0 +1,99 @@
+==============================
+UnsafeRow Serialization Format
+==============================
+
+Velox supports two data serialization formats out of the box:
+`PrestoPage <https://prestodb.io/docs/current/develop/serialized-page.html>`_
+and UnsafeRow. These formats are used in data shuffle. Velox applications
+can register their own formats as well.
+
+PrestoPage format is described in the `Presto documentation <https://prestodb.io/docs/current/develop/serialized-page.html>`_.
+This article describes UnsafeRow format which comes from `Apache Spark <https://jaceklaskowski.gitbooks.io/mastering-spark-sql/content/spark-sql-UnsafeRow.html>`_.
+
+A row is a continuous buffer made of 3 sections:
+
+null bits | fixed-width data | variable-width data
+
+Each section is aligned to 8 bytes.
+
+The 'null bits' section contains one bit per column. 0 bit indicates a
+non-null value. 1 bit indicates a null value.
+
+The 'fixed-width data' section contains 8 bytes per column. Values of the
+fixed-width columns (booleans, integers, floating point numbers) are stored
+directly. These values must fit within 8 bytes. Long decimal columns are not
+supported.
+
+Values of the variable-width columns (strings, arrays, maps) are split between
+fixed-width and variable-width sections. 8 bytes of the fixed-width section
+store the size and location of the value in the variable-width section.
+
+The 'variable-width data' stores values of the variable-width columns. Each
+value is aligned to 8 bytes.
+
+Strings are stored directly without null-terminating byte.
+
+Arrays are stored as 8 bytes for the size of the array followed by array
+elements serialized as UnsafeRow:
+
+array size | null bits | fixed-width data | variable-width data
```

Comment:
I believe for array, it's :

fixed-width data | variable-width data

the fixed-width data is composed of offset (higher 32 bit) and size (lower 32 bit). The first 8 bytes of the variable-width data is representing the number of elements in the array and the second 8 bytes is the null bits

[1] https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java#L123

Reply:
That's right and this is what I have documented. The offset and size are stored in the fixed-width section, which is not necessarily adjacent to the variable-width section. When there are 10 columns, fixed-width section has 80 bytes, 8 bytes per column. Variable-width section starts after fixed-width section and contains only variable-width data.

Reply:
The null bits should be at the head of the variable-width data section?  https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java#L140